File: backend-v5/app/schemas/ops.py
"""Analyst Ops schemas (verify → publish workflow)."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class WorkflowStatus(str, Enum):
    """Analyst workflow status for event clusters."""

    UNREVIEWED = "unreviewed"
    MONITORING = "monitoring"
    VERIFIED = "verified"
    PUBLISHED = "published"
    REJECTED = "rejected"


class OpsStoryItem(BaseModel):
    """Story/evidence item inside an event cluster."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    source_id: str
    source_name: Optional[str] = None
    title: str
    summary: Optional[str] = None
    url: str
    published_at: Optional[datetime] = None


class OpsDuplicateGroup(BaseModel):
    """Group of near-identical stories (soft dedup) within an event."""

    canonical: OpsStoryItem
    duplicates: list[OpsStoryItem] = Field(default_factory=list)

    @property
    def duplicate_count(self) -> int:
        return len(self.duplicates)


class OpsRelatedEvent(BaseModel):
    """Related event suggestion (connect-the-dots)."""

    cluster_id: UUID
    headline: str
    category: Optional[str] = None
    severity: Optional[str] = None
    similarity: float = Field(..., ge=0.0, le=1.0)


class OpsEventInboxItem(BaseModel):
    """Event summary for analyst inbox."""

    id: UUID

    # Display (prefer analyst overrides if present)
    headline: str
    summary: Optional[str] = None
    category: Optional[str] = None
    severity: Optional[str] = None

    # System values (for comparison / audit)
    system_headline: str
    system_category: Optional[str] = None
    system_severity: Optional[str] = None

    story_count: int
    source_count: int
    first_published: Optional[datetime] = None
    last_updated: Optional[datetime] = None

    workflow_status: WorkflowStatus
    is_published: bool
    published_at: Optional[datetime] = None
    verified_at: Optional[datetime] = None

    # Scoring
    age_minutes: Optional[int] = None
    impact_score: float = Field(..., ge=0.0, le=1.0)
    uncertainty_score: float = Field(..., ge=0.0, le=1.0)
    reasons: list[str] = Field(default_factory=list)
    ready_for_publish: bool = False


class OpsEventInboxResponse(BaseModel):
    items: list[OpsEventInboxItem]
    total: int


class OpsEventDetailResponse(BaseModel):
    """Full event details for analyst verification/publishing."""

    id: UUID

    # Display (prefer analyst overrides if present)
    headline: str
    summary: Optional[str] = None
    category: Optional[str] = None
    severity: Optional[str] = None

    # System values
    system_headline: str
    system_summary: Optional[str] = None
    system_category: Optional[str] = None
    system_severity: Optional[str] = None

    story_count: int
    source_count: int
    first_published: Optional[datetime] = None
    last_updated: Optional[datetime] = None

    workflow_status: WorkflowStatus
    analyst_notes: Optional[str] = None
    customer_brief: Optional[str] = None

    is_published: bool
    published_at: Optional[datetime] = None
    verified_at: Optional[datetime] = None

    # Evidence
    story_groups: list[OpsDuplicateGroup]
    all_stories: list[OpsStoryItem]

    # Connect-the-dots
    related_events: list[OpsRelatedEvent] = Field(default_factory=list)


class OpsUpdateEventRequest(BaseModel):
    """Patch analyst fields (does not publish)."""

    analyst_headline: Optional[str] = None
    analyst_summary: Optional[str] = None
    analyst_category: Optional[str] = None
    analyst_severity: Optional[str] = None
    analyst_notes: Optional[str] = None

    workflow_status: Optional[WorkflowStatus] = None


class OpsPublishEventRequest(BaseModel):
    """Publish an event to consumers."""

    customer_brief: Optional[str] = None
    analyst_headline: Optional[str] = None
    analyst_summary: Optional[str] = None
    analyst_category: Optional[str] = None
    analyst_severity: Optional[str] = None

File: frontend/src/api/ops.ts
import apiClient from './client'

export type WorkflowStatus = 'unreviewed' | 'monitoring' | 'verified' | 'published' | 'rejected'

export interface OpsStoryItem {
  id: string
  source_id: string
  source_name?: string | null
  title: string
  summary?: string | null
  url: string
  published_at?: string | null
}

export interface OpsDuplicateGroup {
  canonical: OpsStoryItem
  duplicates: OpsStoryItem[]
}

export interface OpsRelatedEvent {
  cluster_id: string
  headline: string
  category?: string | null
  severity?: string | null
  similarity: number
}

export interface OpsEventInboxItem {
  id: string

  headline: string
  summary?: string | null
  category?: string | null
  severity?: string | null

  system_headline: string
  system_category?: string | null
  system_severity?: string | null

  story_count: number
  source_count: number
  first_published?: string | null
  last_updated?: string | null

  workflow_status: WorkflowStatus
  is_published: boolean
  published_at?: string | null
  verified_at?: string | null

  age_minutes?: number | null
  impact_score: number
  uncertainty_score: number
  reasons: string[]
  ready_for_publish: boolean
}

export interface OpsEventInboxResponse {
  items: OpsEventInboxItem[]
  total: number
}

export interface OpsEventDetailResponse {
  id: string

  headline: string
  summary?: string | null
  category?: string | null
  severity?: string | null

  system_headline: string
  system_summary?: string | null
  system_category?: string | null
  system_severity?: string | null

  story_count: number
  source_count: number
  first_published?: string | null
  last_updated?: string | null

  workflow_status: WorkflowStatus
  analyst_notes?: string | null
  customer_brief?: string | null

  is_published: boolean
  published_at?: string | null
  verified_at?: string | null

  story_groups: OpsDuplicateGroup[]
  all_stories: OpsStoryItem[]

  related_events: OpsRelatedEvent[]
}

export interface OpsUpdateEventRequest {
  analyst_headline?: string | null
  analyst_summary?: string | null
  analyst_category?: string | null
  analyst_severity?: string | null
  analyst_notes?: string | null
  workflow_status?: WorkflowStatus | null
}

export interface OpsPublishEventRequest {
  customer_brief?: string | null
  analyst_headline?: string | null
  analyst_summary?: string | null
  analyst_category?: string | null
  analyst_severity?: string | null
}

export async function getEventInbox(params?: {
  hours?: number
  limit?: number
  minAgeMinutes?: number
  includePublished?: boolean
  needsReviewOnly?: boolean
}): Promise<OpsEventInboxResponse> {
  const response = await apiClient.get('/ops/events/inbox', {
    params: {
      hours: params?.hours ?? 72,
      limit: params?.limit ?? 50,
      min_age_minutes: params?.minAgeMinutes ?? 30,
      include_published: params?.includePublished ?? false,
      needs_review_only: params?.needsReviewOnly ?? false,
    },
  })
  return response.data
}

export async function getEventDetail(clusterId: string): Promise<OpsEventDetailResponse> {
  const response = await apiClient.get(`/ops/events/${clusterId}`)
  return response.data
}

export async function updateEvent(
  clusterId: string,
  body: OpsUpdateEventRequest,
): Promise<OpsEventDetailResponse> {
  const response = await apiClient.patch(`/ops/events/${clusterId}`, body)
  return response.data
}

export async function publishEvent(
  clusterId: string,
  body: OpsPublishEventRequest,
  minAgeMinutes: number = 30,
): Promise<OpsEventDetailResponse> {
  const response = await apiClient.post(`/ops/events/${clusterId}/publish`, body, {
    params: { min_age_minutes: minAgeMinutes },
  })
  return response.data
}

export async function rejectEvent(
  clusterId: string,
  notes?: string,
): Promise<OpsEventDetailResponse> {
  const response = await apiClient.post(`/ops/events/${clusterId}/reject`, null, {
    params: { notes },
  })
  return response.data
}

// ============================================================
// Source Management (Human-in-the-loop)
// ============================================================

export interface StoryManagementResponse {
  status: string
  message: string
  story_id: string
  cluster_id: string
  new_story_count: number
  new_source_count: number
}

/**
 * Remove a story from a cluster without deleting it.
 * The story returns to the unclustered pool.
 */
export async function removeStoryFromCluster(
  clusterId: string,
  storyId: string,
): Promise<StoryManagementResponse> {
  const response = await apiClient.delete(`/ops/events/${clusterId}/stories/${storyId}`)
  return response.data
}

/**
 * Add an unclustered story to an existing cluster.
 */
export async function addStoryToCluster(
  clusterId: string,
  storyId: string,
): Promise<StoryManagementResponse> {
  const response = await apiClient.post(`/ops/events/${clusterId}/stories/${storyId}`)
  return response.data
}

// ============================================================
// Candidate Stories (for adding similar stories)
// ============================================================

export interface CandidateStory {
  story_id: string
  title: string
  url: string
  source_id: string
  source_name?: string | null
  published_at?: string | null
  similarity: number // Percentage (0-100)
  current_cluster_id?: string | null
  current_cluster_headline?: string | null
  is_unclustered: boolean
}

export interface CandidateStoriesResponse {
  cluster_id: string
  cluster_headline: string
  candidates: CandidateStory[]
}

/**
 * Get candidate stories that could be added to a cluster.
 * Returns similar unclustered stories and stories from related clusters.
 */
export async function getCandidateStories(
  clusterId: string,
  limit: number = 20,
): Promise<CandidateStoriesResponse> {
  const response = await apiClient.get(`/ops/events/${clusterId}/candidate-stories`, {
    params: { limit },
  })
  return response.data
}

